### DISCUSSION TITLE
A different javascript Syntax.

### DISCUSSION THREAD

**Discussion by Largo** [ruby/ruby.wasm#455] [Ideas]
This is an explanation of the work that [i have in my fork](https://github.com/largo/ruby.wasm). I'm planning to use this in my RubyKaigi Lightning Talk. I'm currently wondering if I should put this into a separate gem or if we upstream it into the ruby.wasm js package.

I have [Ruby.wasm-quickstart ](https://github.com/Largo/ruby.wasm-quickstart/) and [Videoplayer](https://github.com/Largo/PsychometricStudyVideoPlayer) that use this syntax.

Main difference:

`JS.global.document` instead of `JS.global[:document]`

`button.style.backgroundColor = "#4CAF50"` 
instead of 
`button[:style][:backgroundColor] = "#4CAF50"` 

I also added type conversions for this to work as well as it does.

```ruby
# Update slider position as the video plays
@video_player.addEventListener("timeupdate") do
  @seek_slider.value = @video_player.currentTime unless @seek_slider === @document.activeElement
  @current_time_display.innerText = format_time(@video_player.currentTime)
end

# Handle playback rate adjustments from speed buttons
speed_buttons = @document.querySelectorAll(".speedButton")
speed_buttons.to_a.each do |button|
  button.addEventListener("click") do
    speed_buttons.to_a.each { |btn| btn.style.backgroundColor = "" }
    # Highlight the clicked button
    button.style.backgroundColor = "#4CAF50"

    speed = button.getAttribute("data-speed")
    @video_player.playbackRate = speed.to_f
    @video_player.play
  end
end
```

Any thoughts? :)

## Main blockers
- Splitting up my method_missing is a challenge :)
- I'm struggling with writing tests. (Edit: tests will work after downgrading nodejs to below 20)

**Comment by kateinoigakukun** [comment]
My main concern about the syntax is that `JS.global.document()` seems like a method call but it's property access, and there is no way to distinguish `JS.global.document` and `JS.global.document()` with the current `method_missing`.
My design principle here is not making core things easier but keeping them simpler in the early days. So I'd like to suggest playing it as a gem.
I'll have a complete documentation for RubyGems usage on ruby.wasm soon. Stay tuned.

**Reply by Largo** [reply]
Thank you for your answer. I believe in Ruby properties are philosophically methods, but I understand your concern of what happens if there are a property and a method with the same name. So far I haven't found a real world problem during using this syntax. Maybe we can talk about it during RubyKaigi?

I'll turn it into a gem and show it during my short 5 minute talk :)
My only concern if I make a gem, is how I make sure it chooses the compatible version of the JS gem. I have not done research there yet however.

**Reply by kateinoigakukun** [reply]
Yes, we should talk about this at the Kaigi :)

> My only concern if I make a gem, is how I make sure it chooses the compatible version of the JS gem. I have not done research there yet however.

The js library is now a gem, so you can just specify compatible js gem version in your gemspec.
